File: dat_analyzer.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <cstring>
#include <iomanip>
#include <cstdint>

struct PacketInfo {
    uint8_t type;
    uint8_t source;
    size_t payload_size;
};

uint8_t hexCharToUint8(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
    return 0;
}

void hexDump(const std::vector<uint8_t>& data, const std::string& label, int max_lines = 3) {
    if (!label.empty()) {
        std::cout << "\n" << label << std::endl;
    }
    
    for (size_t i = 0; i < data.size(); i += 16) {
        std::cout << std::hex << std::setfill('0') << std::setw(4) << i << ": ";
        
        // Hex bytes
        for (size_t j = 0; j < 16 && i + j < data.size(); j++) {
            std::cout << std::hex << std::setfill('0') << std::setw(2) << (int)data[i+j] << " ";
        }
        
        // ASCII
        std::cout << "  ";
        for (size_t j = 0; j < 16 && i + j < data.size(); j++) {
            char c = data[i+j];
            std::cout << (c >= 32 && c < 127 ? c : '.');
        }
        std::cout << std::endl;
        
        if (i / 16 >= (size_t)(max_lines - 1)) {
            std::cout << "  ..." << std::endl;
            break;
        }
    }
    std::cout << std::dec;
}

float bytesToFloatLE(const std::vector<uint8_t>& data, size_t offset) {
    if (offset + 3 >= data.size()) return 0.0f;
    uint32_t intVal = (uint32_t)data[offset] | 
                     ((uint32_t)data[offset+1] << 8) | 
                     ((uint32_t)data[offset+2] << 16) | 
                     ((uint32_t)data[offset+3] << 24);
    float floatVal;
    std::memcpy(&floatVal, &intVal, sizeof(float));
    return floatVal;
}

bool analyzePackets(const std::string& filepath, int num_packets) {
    std::ifstream file(filepath);
    if (!file.is_open()) {
        std::cerr << "Error: Could not open " << filepath << std::endl;
        return false;
    }
    
    std::vector<PacketInfo> packets;
    std::map<uint8_t, int> typeCount;
    std::map<uint8_t, int> sourceCount;
    
    std::string line;
    int line_num = 0;
    
    while (std::getline(file, line) && line_num < num_packets) {
        line_num++;
        
        size_t comma_pos = line.find(',');
        if (comma_pos == std::string::npos) continue;
        
        std::string hex_data = line.substr(comma_pos + 1);
        
        if (hex_data.length() < 56) continue;
        
        // Extract source (first byte, chars 0-1)
        uint8_t source = (hexCharToUint8(hex_data[0]) << 4) | hexCharToUint8(hex_data[1]);
        
        // Extract type (byte 20, chars 40-41)
        uint8_t type = (hexCharToUint8(hex_data[40]) << 4) | hexCharToUint8(hex_data[41]);
        
        // Extract payload size
        std::string payload_hex = hex_data.substr(56);
        size_t payload_size = payload_hex.length() / 2;
        
        PacketInfo info;
        info.type = type;
        info.source = source;
        info.payload_size = payload_size;
        
        packets.push_back(info);
        typeCount[type]++;
        sourceCount[source]++;
    }
    
    file.close();
    
    std::cout << "Analyzed " << packets.size() << " packets" << std::endl;
    std::cout << "\nPacket types found:" << std::endl;
    
    for (const auto& pair : typeCount) {
        std::cout << "  Type 0x" << std::hex << std::setfill('0') << std::setw(2) 
                 << (int)pair.first << ": " << std::dec << pair.second << " packets" << std::endl;
    }
    
    std::cout << "\nPacket sources found:" << std::endl;
    for (const auto& pair : sourceCount) {
        std::cout << "  Source 0x" << std::hex << std::setfill('0') << std::setw(2) 
                 << (int)pair.first << ": " << std::dec << pair.second << " packets" << std::endl;
    }
    
    // Detailed analysis by type
    std::cout << "\n\n=== DETAILED PACKET ANALYSIS ===" << std::endl;
    
    // Type 0x07
    if (typeCount.count(0x07)) {
        std::cout << "\n--- TYPE 0x07 (Own Trajectory) ---" << std::endl;
        file.open(filepath);
        int count = 0;
        while (std::getline(file, line) && count < 2) {
            size_t comma_pos = line.find(',');
            if (comma_pos == std::string::npos) continue;
            
            std::string hex_data = line.substr(comma_pos + 1);
            if (hex_data.length() < 42) continue;
            
            if (hex_data.substr(40, 2) != "07") continue;
            
            std::string timestamp = line.substr(0, comma_pos);
            std::string payload_hex = hex_data.substr(56);
            
            std::vector<uint8_t> payload;
            for (size_t i = 0; i < payload_hex.length(); i += 2) {
                if (i + 1 < payload_hex.length()) {
                    uint8_t byte = (hexCharToUint8(payload_hex[i]) << 4) | 
                                  hexCharToUint8(payload_hex[i+1]);
                    payload.push_back(byte);
                }
            }
            
            std::cout << "\nPacket #" << (count + 1) << ": " << timestamp << std::endl;
            std::cout << "Payload size: " << payload.size() << " bytes" << std::endl;
            
            hexDump(payload, "Payload (hex):");
            
            std::cout << "\nAs little-endian u32/f32:" << std::endl;
            for (size_t i = 0; i < 8 && i * 4 + 4 <= payload.size(); i++) {
                uint32_t u32_val = (uint32_t)payload[i*4] | 
                                  ((uint32_t)payload[i*4+1] << 8) | 
                                  ((uint32_t)payload[i*4+2] << 16) | 
                                  ((uint32_t)payload[i*4+3] << 24);
                float f32_val = bytesToFloatLE(payload, i*4);
                
                std::cout << "  Offset " << std::setw(2) << (i*4) << ": u32=" 
                         << std::setw(10) << u32_val << "  f32=" 
                         << std::fixed << std::setprecision(2) << std::setw(10) << f32_val << std::endl;
            }
            
            count++;
        }
        file.close();
    }
    
    // Type 0x0d
    if (typeCount.count(0x0d)) {
        std::cout << "\n--- TYPE 0x0d (Target/Torpedo) ---" << std::endl;
        file.open(filepath);
        int count = 0;
        while (std::getline(file, line) && count < 1) {
            size_t comma_pos = line.find(',');
            if (comma_pos == std::string::npos) continue;
            
            std::string hex_data = line.substr(comma_pos + 1);
            if (hex_data.length() < 42) continue;
            
            if (hex_data.substr(40, 2) != "0d") continue;
            
            std::string timestamp = line.substr(0, comma_pos);
            std::string payload_hex = hex_data.substr(56);
            
            std::vector<uint8_t> payload;
            for (size_t i = 0; i < payload_hex.length(); i += 2) {
                if (i + 1 < payload_hex.length()) {
                    uint8_t byte = (hexCharToUint8(payload_hex[i]) << 4) | 
                                  hexCharToUint8(payload_hex[i+1]);
                    payload.push_back(byte);
                }
            }
            
            std::cout << "\nPacket #" << (count + 1) << ": " << timestamp << std::endl;
            std::cout << "Payload size: " << payload.size() << " bytes" << std::endl;
            
            hexDump(payload, "Payload (hex):", 5);
            
            std::cout << "\nAs little-endian u32/f32 (first 16 fields):" << std::endl;
            for (size_t i = 0; i < 16 && i * 4 + 4 <= payload.size(); i++) {
                uint32_t u32_val = (uint32_t)payload[i*4] | 
                                  ((uint32_t)payload[i*4+1] << 8) | 
                                  ((uint32_t)payload[i*4+2] << 16) | 
                                  ((uint32_t)payload[i*4+3] << 24);
                float f32_val = bytesToFloatLE(payload, i*4);
                
                std::cout << "  Offset " << std::setw(2) << (i*4) << ": u32=" 
                         << std::setw(10) << u32_val << "  f32=" 
                         << std::fixed << std::setprecision(2) << std::setw(10) << f32_val << std::endl;
            }
            
            count++;
        }
        file.close();
    }
    
    return true;
}

int main(int argc, char* argv[]) {
    std::string filepath = "/home/siddartha/Music/FiringAnalysis/saveData.dat";
    int num_packets = 30;
    
    if (argc > 1) {
        filepath = argv[1];
    }
    if (argc > 2) {
        num_packets = std::stoi(argv[2]);
    }
    
    std::cout << "DatAnalyzer - Binary .dat file analyzer" << std::endl;
    std::cout << "File: " << filepath << std::endl;
    std::cout << "Analyzing first " << num_packets << " packets..." << std::endl;
    
    if (!analyzePackets(filepath, num_packets)) {
        return 1;
    }
    
    return 0;
}

File: dat_analyzer.py
#!/usr/bin/env python3
"""
Utilities to analyze and decode .dat file packets
"""
import struct
import sys
import re
from collections import defaultdict

def hex_dump(data, label=""):
    """Print a formatted hex dump of binary data"""
    if label:
        print(f"\n{label}")
    for i in range(0, len(data), 16):
        chunk = data[i:i+16]
        hex_str = ' '.join(f'{b:02x}' for b in chunk)
        ascii_str = ''.join(chr(b) if 32 <= b < 127 else '.' for b in chunk)
        print(f"{i:04x}: {hex_str:<48} {ascii_str}")

def parse_dat_file(filepath, num_packets=10):
    """Parse first N packets from .dat file"""
    packets = []
    
    with open(filepath, 'r') as f:
        for line_num, line in enumerate(f):
            if line_num >= num_packets:
                break
            
            line = line.strip()
            if not line:
                continue
                
            # Parse: timestamp,hexdata
            parts = line.split(',')
            if len(parts) < 2:
                continue
            
            timestamp = parts[0]
            hex_data = parts[1]
            
            #Extract type from fixed offset (byte 20)
            if len(hex_data) >= 42:
                type_str = hex_data[40:42]
                packet_type = int(type_str, 16)
            else:
                packet_type = 0
            
            payload_hex = hex_data[56:]  # Everything after 28-byte header
            payload_bytes = bytes.fromhex(payload_hex)
            
            packets.append({
                'timestamp': timestamp,
                'hex_header': hex_data[:56],
                'hex_payload': payload_hex,
                'type': packet_type,
                'payload': payload_bytes,
                'line': line_num
            })
    
    return packets

def analyze_packets(filepath, num_packets=30):
    """Analyze multiple packets to find patterns"""
    packets = parse_dat_file(filepath, num_packets)
    
    # Group by type
    by_type = defaultdict(list)
    for pkt in packets:
        by_type[pkt['type']].append(pkt)
    
    print(f"Analyzed {len(packets)} packets")
    print(f"\nPacket types found:")
    for ptype in sorted(by_type.keys()):
        print(f"  Type 0x{ptype:02x}: {len(by_type[ptype])} packets")
    
    # Analyze type 0x07 (Own trajectory)
    if 0x07 in by_type:
        print(f"\n\n=== TYPE 0x07 (Own Trajectory) ===")
        for pkt in by_type[0x07][:3]:  # First 3 examples
            print(f"\nPacket #{pkt['line']}: {pkt['timestamp']}")
            print(f"Payload size: {len(pkt['payload'])} bytes")
            
            hex_dump(pkt['payload'][:60], "Payload (hex):")
            
            # Try various interpretations
            print("\nAs little-endian u32/f32:")
            for i in range(0, min(len(pkt['payload']), 32), 4):
                if i + 4 <= len(pkt['payload']):
                    u32 = struct.unpack('<I', pkt['payload'][i:i+4])[0]
                    f32 = struct.unpack('<f', pkt['payload'][i:i+4])[0]
                    print(f"  Offset {i:2d}: u32={u32:10d}  f32={f32:10.2f}")
    
    # Analyze type 0x0d (Target/Torpedo)  
    if 0x0d in by_type:
        print(f"\n\n=== TYPE 0x0d (Target/Torpedo) ===")
        for pkt in by_type[0x0d][:1]:  # First example
            print(f"\nPacket #{pkt['line']}: {pkt['timestamp']}")
            print(f"Payload size: {len(pkt['payload'])} bytes")
            
            hex_dump(pkt['payload'][:80], "Payload (hex):")
            
            print("\nAs little-endian u32/f32:")
            for i in range(0, min(len(pkt['payload']), 64), 4):
                if i + 4 <= len(pkt['payload']):
                    u32 = struct.unpack('<I', pkt['payload'][i:i+4])[0]
                    f32 = struct.unpack('<f', pkt['payload'][i:i+4])[0]
                    print(f"  Offset {i:2d}: u32={u32:10d}  f32={f32:10.2f}")

if __name__ == '__main__':
    analyze_packets('/home/siddartha/Music/FiringAnalysis/saveData.dat', 30)
